chore(release): release v0.10.0 (#1782)
qri-io/qri: CHANGELOG.md

## Massive Improvements to Save performance
We've drastically improved the reliability and scalability of saving a dataset on Qri. Qri uses a bounded block of memory while saving, meaning it will only consume roughly a MAX of 150MB of memory while saving, regardless of how large your dataset is. This means the max size of dataset you can save is no longer tied to your available memory.

We've had to change some underlying functionality to get the scalability we want, to that end we no longer calculate `Structure.Checksum`, we no longer calculate commit messages for datasets over a certain size, and we no longer store all the error values found when validating the body of a dataset.

## API Overhaul

> Our biggest change has been a complete overhaul of our API.
> 
> Except for specific "sugar" endpoints, which can be fetched using "GET" requests, our entire api uses JSON HTTP "POST" requests, with the request body reflecting the associated parameters of that `lib` method.

Reviewer: ```suggestion
Our biggest change has been a complete overhaul of our API.

Except for specific "sugar" endpoints, which can be fetched using "GET" requests, our entire api uses JSON HTTP "POST" requests, with the request body reflecting the associated parameters of that `lib` method.

We wanted to make our API easier to work with by making it more consistent across endpoints. After a great deal of review & discussion, this overhaul introduces an RPC-style centric API that expects JSON `POST` requests, plus a few `GET` requests we're calling "sugar" endpoints.

The RPC part of our api is an HTTP pass-through to our lib methods. This makes working with qri over HTTP the _same_ as working with Qri as a library. We've spent a lot of time building & organizing qri's `lib` interface, and now all of that same functionality is exposed over HTTP. The intended audience for the RPC API are folks who want to automate qri across process boundaries, and still have very fine grained control. Think "command line over HTTP".

At the same time, however, we didn't want to lose a number of important-to-have endpoints, like being able to `GET` a dataset body via just a URL string, so we've moved all of these into a "sugar" API, and made lots of room to grow. We'll continue to add convenience-oriented endpoints that make it easy to work with Qri. The "sugar" API will be oriented to users who are prioritizing fetching data from Qri to use elsewhere.
```

| "/sync/dag"                              | GET *2              | []byte           | remote.DsyncHTTPHandler          |
| "/sync/dag"                              | POST                | dag.Manifest     | remote.DsyncHTTPHandler          |
| "/sync/dag"                              | PUT                 | --               | remote.DsyncHTTPHandler          |
| "/sync/dag"                              | PATCH               | stream?          | remote.DsyncHTTPHandler          |
| "/sync/logs"                             | GET                 | --               | remote.LogbookHTTPHandlery       |
| "/sync/logs"                             | PUT                 | []byte           | remote.LogbookHTTPHandlery       |
| "/sync/logs"                             | DELETE              | --               | remote.LogbookHTTPHandlery       |
| "/sync/resolve"                          | GET                 | DSRef            | remote.RefsHTTPHandler           |
| "/sync/resolve"                          | DELETE              | --               | remote.RefsHTTPHandler           |

> |                                          |                     |                  |                                  |

Reviewer: let's remove this section on protocol endpoints, we don't want users calling these.

## Redesigned lib interface

> In general, we've streamlined the core functionality and reconciled input params in the `lib` package (which contain the methods and params that power both the api and cmd), so that no matter how you are accessing functionality in qri, whether using `lib` as a package,  using the api, or using the command line, you will have, not only consistent input parameters, but also consistent behavior. We're also utilizing our new `dispatch` pattern to replace our old rpc client with the same http json api that we expose to our users. That way all our api's, http or otherwise, have the same expectations. If something is broke in one place, it is broken in all places, consequently, when it is fixed in one place it will be fixed in all places.

Reviewer: ```suggestion
In general, we've streamlined the core functionality and reconciled input params in the `lib` package (which contain the methods and params that power both the api and cmd), so that no matter how you are accessing functionality in qri, whether using `lib` as a package,  using the HTTP API, or using the command line, you can expect consistent inputs and–more importantly–consistent behavior. We're also utilizing our new `dispatch` pattern to replace our old rpc client with the _same_ JSON HTTP API exposed to users. That way all our API's, HTTP or otherwise have the same expectations. If something is broke in one place, it is broken in all places, consequently, when it is fixed in one place it will be fixed in all places!
```

These changes will also help us in our upcoming challenges to refine expand the notion of identity inside of Qri.

## Step-Based Transform Execution

> We've changed the underlying way that transforms execute, dividing the work into steps, like `Setup`, `Download`, and `Transform`. With this change, we are able to report the status of their transform back to the user. It will also give the user a better sense of what is happening in their transform: how long did the `Download` step take, which step failed, which steps worked. It also paves the way for a better transform development environment in the future. Users will be able to re-run certain steps or cache steps that they don't want to waste time re-running.

Reviewer: I don't think this feature is quite ready to talk about. Yes, it's in Qri, but at this point I'd label it _pre-experimental_. Let's remove this, and save this exact paragraph for the next set of release notes

## Other features:
### Change report over API

> We've added an endpoint to get the change report via the api: `/changes`. This lays the ground work for powering any future UI we have surrounding change reports in our frontend applications.

Reviewer: ```suggestion
We've added an endpoint to get the change report via the api: `/changes`. Laying the groundwork for future User Interfaces that can report on changes between versions.
```

### Access and Oauth

> In preparation for multi tenency, we are adding support for generating JSON web tokens that will help ensure identity on the network. You can generate an access token via the cmd using the `qri access` command, or over the api using the `/access/token` endpoint.

Reviewer: ```suggestion
We're working towards making qri-core able to handle multiple users on the same node (AKA multi-tenancy) In preparation for multi-tenancy, we are adding support for generating JSON web tokens (JWTs) that will help ensure identity on the network. You can generate an access token via the cmd using the `qri access` command, or over the api using the `/access/token` endpoint.
```

> ### Stats Component!
> We've added a new component: Stats! Stats is a component that contains statistical metadata about the body of a dataset. Its purpose is to provide an "at a glance" summary of a dataset. In order to remain fast for very large dataset sizes, we have opted to calculate the stats using probabilistic structures. Stats are an important part of change reports, and allow you to get a sense of what is different in a dataset body without having to examine that body line by line. 

Reviewer: From a user's perspective this is actually a very big deal. I'd move this up to just after the `Resdesigned Lib Interface` section.

```suggestion
### Stats Component!
We've added a new component: Stats! Stats is a component that contains statistical metadata about the body of a dataset. Stats are now automatically calculated and saved with each new version.

Its purpose is to provide an "at a glance" summary of a dataset, calculating statistics on columns (ok, on "compound data types", but it's much easier to think about column stats). In order to remain fast for very large dataset sizes, we have opted to calculate the stats using probabilistic structures. Stats are an important part of change reports, and allow you to get a sense of what is different in a dataset body without having to examine that body line by line. 

For earlier versions that don't have stats calculated, we've added a `qri stats` command that will calculate the new stats component for you.
```

### Apply Command
You can now use the `qri apply` command to "dry run" and de-bug transforms! Use the `--file` flag to apply a new `transform.star` file, use a dataset reference to re-run an already existing transform on an already existing dataset, or use both to apply a new tranform to an already existing dataset. The resulting dataset is output to the terminal. To run the transform and save the results as a new commit, use the `--apply` flag in the `qri save` command.

### Progress bars
Qri has been fine tuned to handle larger datasets faster. Regardless of how fast your dataset is saving, we want to be able to track any progress. Now, when saving a new dataset or a new version of a dataset, the command line will show you a progress bar.

> ### `qri version`
> To help with debugging, the `qri version` command now comes chock full of details, including the qri version, the build time, and the git summary.

Reviewer: ```suggestion
### `qri version`
To help with debugging, the `qri version` command now comes chock full of details, including the qri version, the build time, and the git summary. (developers: you need to use `make build` now instead of `go build`!)
```
